src/client/portforwarder.go

```diff
@@ -0,0 +1,191 @@
+package client
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"path"
+	"os"
+
+	"github.com/facebookgo/pidfile"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	daemonLocalPort = 30650
+	samlAcsLocalPort = 30654
+	dashUILocalPort = 30080
+	dashWebSocketLocalPort = 30081
+)
+
+// PortForwarder handles proxying local traffic to a kubernetes pod
+type PortForwarder struct {
+	core corev1.CoreV1Interface
+	client rest.Interface
+	config *rest.Config
+	namespace string
+	stdout io.Writer
+	stderr io.Writer
+	stopChansLock *sync.Mutex
+	stopChans []chan struct{}
+	shutdown bool
+}
+
+// NewPortForwarder creates a new port forwarder
+func NewPortForwarder(namespace string, stdout, stderr io.Writer) (*PortForwarder, error) {
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	rules := clientcmd.NewDefaultClientConfigLoadingRules()
+	overrides := &clientcmd.ConfigOverrides{}
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(rules, overrides)
+	config, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	core := client.CoreV1()
+
+	return &PortForwarder {
+		core: core,
+		client: core.RESTClient(),
+		config: config,
+		namespace: namespace,
+		stdout: stdout,
+		stderr: stderr,
+		stopChansLock: &sync.Mutex{},
+		stopChans: []chan struct{}{},
+		shutdown: false,
+	}, nil
+}
+
+// Run starts the port forwarder. Returns after initialization is begun,
+// returning any initialization errors.
+func (f *PortForwarder) Run(podNameSelector map[string]string, localPort, remotePort int) error {
+	podList, err := f.core.Pods(f.namespace).List(metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(podNameSelector)),
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ListOptions",
+			APIVersion: "v1",
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if len(podList.Items) != 1 {
```

Comment:
This probably shouldn't error. Pachd is designed so that you can have more than one copy of it running at a time and use any of the instances as endpoints. We should probably just select a random item from the slice, and only error if it's empty.

Reply:
What about just choosing the first item instead? Random seems like it could cause confusing behavior.

Reply:
I don't think it'll be confusing. You should get exactly the same semantics if you're sending multiple commands to different `pachd` endpoints as if you were sending them to the same endpoint. Random is pretty close to the semantics you'll get if you connect through a service (stateful round robin instead of stateless random).